[Installation]: building docker cpu image with VLLM_CPU_DISABLE_AVX512=true (or on any x86_64 CPU without AVX512) fails to compile mla_decode.cpp because BFloat16 has no AVX2 fallback

### Your current environment

```text
Collecting environment information...
uv is set
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.3 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version                : 18.1.3 (1ubuntu1)
CMake version                : version 3.30.2
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.9.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.10 (main, Apr  9 2025, 04:03:51) [Clang 20.1.0 ] (64-bit runtime)
Python platform              : Linux-6.8.0-90-generic-x86_64-with-glibc2.39

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.1.115
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA TITAN RTX
Nvidia driver version        : 590.48.01
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i9-10900K CPU @ 3.70GHz
CPU family:                           6
Model:                                165
Thread(s) per core:                   2
Core(s) per socket:                   10
Socket(s):                            1
Stepping:                             5
CPU(s) scaling MHz:                   86%
CPU max MHz:                          5300.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7399.70
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp pku ospke md_clear flush_l1d arch_capabilities ibpb_exit_to_user
L1d cache:                            320 KiB (10 instances)
L1i cache:                            320 KiB (10 instances)
L2 cache:                             2.5 MiB (10 instances)
L3 cache:                             20 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected
Vulnerability Vmscape:                Mitigation; IBPB before exit to userspace

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.3
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.16.0
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.3.3
[pip3] nvidia-ml-py==13.590.44
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.3.20
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.9.0
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.24.0
[pip3] transformers==4.57.3
[pip3] triton==3.5.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.12.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	GPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	0-19	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
```


### How you are installing vllm

```sh
docker build -f docker/Dockerfile.cpi \
        --tag vllm-cpu-env \
        --target vllm-openai .
```

```sh
docker build -f docker/Dockerfile.cpi \
        --build-arg VLLM_CPU_DISABLE_AVX512=true \
        --tag vllm-cpu-env \
        --target vllm-openai .
```

My cpu lacks the avx512 instruction set and as such I cannot test what this build would be like on a machine that has the instruction set. For me both of the above builds fail with the following error message:

```
 => ERROR [vllm-build 6/6] RUN --mount=type=cache,target=/root/.cache/uv     --mount=type=cache,target=/root/.cache/ccache     --mount=type=cache,target=/vllm-workspace/.deps,sharing=locked     VLLM_TARGET_DEVICE=cpu python3  19.3s
------
 > [vllm-build 6/6] RUN --mount=type=cache,target=/root/.cache/uv     --mount=type=cache,target=/root/.cache/ccache     --mount=type=cache,target=/vllm-workspace/.deps,sharing=locked     VLLM_TARGET_DEVICE=cpu python3 setup.py bdist_wheel --dist-dir=dist --py-limited-api=cp38:
3.528 /opt/venv/lib/python3.12/site-packages/setuptools_scm/_integration/version_inference.py:51: UserWarning: version of None already set
3.528   warnings.warn(self.message)
... truncated to fit maximum field length ...
4.226 Using MAX_JOBS=32 as the number of jobs.
4.393 -- The CXX compiler identification is GNU 12.3.0
4.402 -- Detecting CXX compiler ABI info
4.454 -- Detecting CXX compiler ABI info - done
4.466 -- Check for working CXX compiler: /usr/bin/g++-12 - skipped
4.467 -- Detecting CXX compile features
4.467 -- Detecting CXX compile features - done
4.481 -- Build type: RelWithDebInfo
4.481 -- Target device: cpu
4.598 -- Found Python: /opt/venv/bin/python3 (found version "3.12.12") found components: Interpreter Development.Module Development.SABIModule
4.598 -- Found python matching: /opt/venv/bin/python3.
5.625 CMake Warning at /opt/venv/lib/python3.12/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:22 (message):
5.625   static library kineto_LIBRARY-NOTFOUND not found.
5.625 Call Stack (most recent call first):
5.625   /opt/venv/lib/python3.12/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:125 (append_torchlib_if_found)
5.625   CMakeLists.txt:91 (find_package)
5.625
5.625
5.625 -- Found Torch: /opt/venv/lib/python3.12/site-packages/torch/lib/libtorch.so
5.632 -- CPU extension compile flags: -mf16c;-fopenmp;-DVLLM_CPU_EXTENSION;-mavx2
5.632 -- Generating CPU attention dispatch header
5.632 CMake Warning at cmake/cpu_extension.cmake:175 (message):
5.632   vLLM CPU backend using AVX2 ISA
5.632 Call Stack (most recent call first):
5.632   CMakeLists.txt:111 (include)
5.632
5.632
5.657 -- CPU extension source files: csrc/cpu/activation.cpp;csrc/cpu/utils.cpp;csrc/cpu/layernorm.cpp;csrc/cpu/mla_decode.cpp;csrc/cpu/pos_encoding.cpp;csrc/moe/dynamic_4bit_int_moe_cpu.cpp;csrc/cpu/cpu_attn.cpp;csrc/cpu/torch_bindings.cpp
5.673 -- Enabling C extension.
5.673 -- Configuring done (1.3s)
5.683 -- Generating done (0.0s)
5.683 -- Build files have been written to: /vllm-workspace/build/temp.linux-x86_64-cpython-312
5.686 Using MAX_JOBS=32 as the number of jobs.
5.880 [1/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/activation.cpp.o
5.894 [2/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/utils.cpp.o
5.900 [3/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/pos_encoding.cpp.o
5.906 [4/9] Building CXX object CMakeFiles/_C.dir/csrc/moe/dynamic_4bit_int_moe_cpu.cpp.o
5.912 [5/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/torch_bindings.cpp.o
5.918 [6/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/layernorm.cpp.o
6.711 [7/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/cpu_attn.cpp.o
18.93 [8/9] Building CXX object CMakeFiles/_C.dir/csrc/cpu/mla_decode.cpp.o
18.93 FAILED: [code=1] CMakeFiles/_C.dir/csrc/cpu/mla_decode.cpp.o
18.93 ccache /usr/bin/g++-12 -DPy_LIMITED_API=3 -DTORCH_EXTENSION_NAME=_C -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -D_C_EXPORTS -I/vllm-workspace/csrc -isystem /opt/uv/python/cpython-3.12-linux-x86_64-gnu/include/python3.12 -isystem /opt/venv/lib/python3.12/site-packages/torch/include -isystem /opt/venv/lib/python3.12/site-packages/torch/include/torch/csrc/api/include -O2 -g -DNDEBUG -std=gnu++17 -fPIC -mf16c -fopenmp -DVLLM_CPU_EXTENSION -mavx2 -MD -MT CMakeFiles/_C.dir/csrc/cpu/mla_decode.cpp.o -MF CMakeFiles/_C.dir/csrc/cpu/mla_decode.cpp.o.d -o CMakeFiles/_C.dir/csrc/cpu/mla_decode.cpp.o -c /vllm-workspace/csrc/cpu/mla_decode.cpp
18.93 /vllm-workspace/csrc/cpu/mla_decode.cpp: In instantiation of ‘void mla_decode_kvcache_cpu_impl(scalar_t*, const scalar_t*, const scalar_t*, int, float, const int*, const int*, int, int, int, int, int) [with scalar_t = c10::BFloat16; int HEAD_DIM = 576; int V_HEAD_DIM = 512; int BLOCK_SIZE = 16]’:
18.93 /vllm-workspace/csrc/cpu/mla_decode.cpp:387:3:   required from here
18.93 /vllm-workspace/csrc/cpu/mla_decode.cpp:268:44: error: incomplete type ‘qk_vec_type’ {aka ‘void’} used in nested name specifier
18.93   268 |   constexpr int QK_NUM_ELEM = qk_vec_type::VEC_ELEM_NUM;
18.93       |                                            ^~~~~~~~~~~~
18.93 ninja: build stopped: subcommand failed.
18.93 Traceback (most recent call last):
18.93   File "/vllm-workspace/setup.py", line 1030, in <module>
18.93     setup(
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/__init__.py", line 117, in setup
18.93     return distutils.core.setup(**attrs)
18.93            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 186, in setup
18.93     return run_commands(dist)
18.93            ^^^^^^^^^^^^^^^^^^
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
18.93     dist.run_commands()
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1002, in run_commands
18.93     self.run_command(cmd)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/dist.py", line 1105, in run_command
18.93     super().run_command(command)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
18.93     cmd_obj.run()
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/command/bdist_wheel.py", line 370, in run
18.93     self.run_command("build")
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 357, in run_command
18.93     self.distribution.run_command(command)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/dist.py", line 1105, in run_command
18.93     super().run_command(command)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
18.93     cmd_obj.run()
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/command/build.py", line 135, in run
18.93     self.run_command(cmd_name)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 357, in run_command
18.93     self.distribution.run_command(command)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/dist.py", line 1105, in run_command
18.93     super().run_command(command)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
18.93     cmd_obj.run()
18.93   File "/vllm-workspace/setup.py", line 360, in run
18.93     super().run()
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/command/build_ext.py", line 99, in run
18.93     _build_ext.run(self)
18.93   File "/opt/venv/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 368, in run
18.93     self.build_extensions()
18.93   File "/vllm-workspace/setup.py", line 329, in build_extensions
18.93     subprocess.check_call(["cmake", *build_args], cwd=self.build_temp)
18.93   File "/opt/uv/python/cpython-3.12.12-linux-x86_64-gnu/lib/python3.12/subprocess.py", line 413, in check_call
18.93     raise CalledProcessError(retcode, cmd)
18.93 subprocess.CalledProcessError: Command '['cmake', '--build', '.', '-j=32', '--target=_C']' returned non-zero exit status 1.
------
Dockerfile.cpu:127
--------------------
 126 |
 127 | >>> RUN --mount=type=cache,target=/root/.cache/uv \
 128 | >>>     --mount=type=cache,target=/root/.cache/ccache \
 129 | >>>     --mount=type=cache,target=/vllm-workspace/.deps,sharing=locked \
 130 | >>>     VLLM_TARGET_DEVICE=cpu python3 setup.py bdist_wheel --dist-dir=dist --py-limited-api=cp38
 131 |
--------------------
ERROR: failed to solve: process "/bin/sh -c VLLM_TARGET_DEVICE=cpu python3 setup.py bdist_wheel --dist-dir=dist --py-limited-api=cp38" did not complete successfully: exit code: 1
```

(the line count is slightly off from the reference dockerfile because I have added the following lines:
```
# MATT OVERRIDE
ARG SETUPTOOLS_SCM_PRETEND_VERSION
ENV SETUPTOOLS_SCM_PRETEND_VERSION=${SETUPTOOLS_SCM_PRETEND_VERSION:-}
```
starting line 111 to workaround #33987 

## AI GENERATED SUGGESTION

I asked claude to diagnose this issue and it stated the following (light editing to reduce repetition):

> The error is that qk_vec_type resolves to void for BFloat16 with AVX2-only (no AVX512). The MLA decode kernel has SIMD specializations that are missing the AVX2+BFloat16 combination. Let me look at the source.
>
> Look at lines 34-57: KernelVecType<c10::BFloat16> is only specialized for __AVX512BF16__, __s390x__, and __aarch64__. On x86_64 without AVX512, there's no specialization for BFloat16, so it falls through to the base template where qk_vec_type = void.
>
> This is a bug in vLLM — the MLA decode kernel is missing an x86_64 AVX2 fallback for BFloat16. You should file this as a separate issue. The fix would be adding an #else branch for generic x86_64 (after the __aarch64__ block, before #endif) that uses FP32 vectors as the fallback, similar to what s390x and aarch64 do. However, this requires checking that BF16Vec16 is actually available on AVX2. Let me verify.
>
> ...
>
> BF16Vec16 on x86 uses __m256i (AVX2 256-bit registers), so it's available without AVX512. The fallback is safe.
>
> The problem: mla_decode.cpp lines 34-57 have #ifdef chains for BFloat16 specialization covering __AVX512BF16__, __s390x__, and __aarch64__, but no #else fallback for x86_64 with only AVX2. So KernelVecType<c10::BFloat16> resolves to the base template with qk_vec_type = void, causing the compile error.
>
> The fix: Change the #endif on line 57 to an #else with an AVX2 fallback, matching the pattern used by s390x/aarch64 (convert BF16 to FP32 for compute):

```
#elif defined(__aarch64__)
template <>
struct KernelVecType<c10::BFloat16> {
  using qk_load_vec_type = vec_op::BF16Vec16;
  using qk_vec_type = vec_op::FP32Vec16;
  using v_load_vec_type = vec_op::BF16Vec16;
};
#else
// x86_64 without AVX512BF16: load as BF16, compute as FP32
template <>
struct KernelVecType<c10::BFloat16> {
  using qk_load_vec_type = vec_op::BF16Vec16;
  using qk_vec_type = vec_op::FP32Vec16;
  using v_load_vec_type = vec_op::BF16Vec16;
};
#endif
```

> You should file this as a bug in vLLM — building with VLLM_CPU_DISABLE_AVX512=true (or on any x86_64 CPU without AVX512) fails to compile mla_decode.cpp because BFloat16 has no AVX2 fallback. It's the same fix pattern already used for aarch64 and s390x.

I am not proficient in C++ so I cannot comment on the proposed fix. I will say that with it the vllm docker image builds and I am able to run Qwen/Qwen3-0.6B successfully. There is a related issue for arm/amd64 [here](https://github.com/vllm-project/vllm/issues/33784)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

--- a/csrc/cpu/mla_decode.cpp
+++ b/csrc/cpu/mla_decode.cpp
@@ -38,16 +38,7 @@ struct KernelVecType<c10::BFloat16> {
   using qk_vec_type = vec_op::BF16Vec32;
   using v_load_vec_type = vec_op::BF16Vec16;
 };
-
-#elif defined(__s390x__)
-template <>
-struct KernelVecType<c10::BFloat16> {
-  using qk_load_vec_type = vec_op::BF16Vec16;
-  using qk_vec_type = vec_op::FP32Vec16;
-  using v_load_vec_type = vec_op::BF16Vec16;
-};
-
-#elif defined(__aarch64__)
+#else
 template <>
 struct KernelVecType<c10::BFloat16> {
   using qk_load_vec_type = vec_op::BF16Vec16;

